File: Round 1B/noisy-neighbors.cpp
// Copyright (c) 2015 kamyu. All rights reserved.

/*
 * UVa10340 - All in All
 * https://code.google.com/codejam/contest/8224486/dashboard#s=p1
 *
 * Time  : O(m), m is length of t.
 * Space : O(1)
 *
 */

#include <iostream>
#include <vector>
#include <utility>
#include <algorithm>
#include <functional>
#include <numeric>

using std::cin;
using std::cout;
using std::endl;
using std::vector;
using std::pair;
using std::greater;
using std::max;

int main() {
    int T;
    cin >> T;

    for (int test = 1; test <= T; ++test) {
        int R, C, N; cin >> R >> C >> N;

        int answer = 0;
        if (N > (R * C + 1) / 2) {
            const int vacant_num = R * C - N;
            vector<pair<int, int>> directions{{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
            vector<int> edges[2];
            for (int i = 0; i < R; ++i) {
                for (int j = 0; j < C; ++j) {
                    int count = 0;
                    for (const auto& d : directions) {
                        if (i + d.first < 0 || i + d.first >= R ||
                            j + d.second < 0 || j + d.second >= C) {
                            continue;
                        }
                        ++count;
                    }
                    edges[(i + j) % 2].push_back(count);
                }
            }
            sort(edges[0].begin(), edges[0].end(), greater<int>());
            sort(edges[1].begin(), edges[1].end(), greater<int>());

            // Max edges of even rooms.
            int even_edges = accumulate(edges[0].cbegin(),
                                        edges[0].cbegin() + vacant_num, 0);
            // Max edges of odd rooms.
            int odd_edges = accumulate(edges[1].cbegin(),
                                       edges[1].cbegin() + vacant_num, 0);
            // Total edges - max edges of vacant rooms.
            answer = R * (C - 1) + (R - 1) * C - max(even_edges, odd_edges);
        }
        cout << "Case #" << test << ": " << answer << endl;
    }
}

File: Round 1B/noisy-neighbors.py
# Copyright (c) 2015 kamyu. All rights reserved.
#
# Google Code Jam 2015 Round 1B - Problem B. Noisy Neighbors
# https://code.google.com/codejam/contest/8224486/dashboard#s=p1
#
# Time:  O(R * C)
# Space: O(1)
#
# It does exist Time: O(1) solution, but the formula is error-prone, 
# hard to figure out it during the contest time.
#

def unhappiness(R, C, N):
    answer = 0
    if N > (R * C + 1) / 2:
        vacant_num = R * C - N
        directions = [[0, 1], [0, -1], [1, 0], [-1, 0]]
        # Number of rooms for each edge number.
        rooms = [[0 for _ in xrange(5)] for _ in xrange(2)]
        for i in xrange(R):
            for j in xrange(C):
                count = 0
                for d in directions:
                    if i + d[0] < 0 or i + d[0] >= R or \
                        j + d[1] < 0 or j + d[1] >= C:
                        continue
                    count += 1
                rooms[(i+j)%2][count] += 1

        # Max edge number of vacant even rooms.
        even_edges, remaining = 0, vacant_num
        for count in reversed(xrange(5)):
            even_edges += count * min(rooms[0][count], remaining)
            remaining -= min(rooms[0][count], remaining)
        
        # Max edge number of vacant odd rooms.
        odd_edges, remaining = 0, vacant_num
        for count in reversed(xrange(5)):
            odd_edges += count * min(rooms[1][count], remaining)
            remaining -= min(rooms[1][count], remaining)

        # (total edge number) - (max edge numbers of vacant rooms).
        answer = (R * (C - 1) + (R - 1) * C) - max(even_edges, odd_edges)
    return answer

for case in xrange(input()):
    R, C, N = map(int, raw_input().strip().split())
    print("Case #%d: %d" % (case+1, unhappiness(R, C, N)))
